### PR TITLE
chase: handle ".." more carefully

### PR DESCRIPTION
**Pull request by yuwata**
This adds several assertions, to clarify what causes #28458.

### CODE REVIEW

File: src/basic/chase.c [hunk 1]
```diff
@@ -212,30 +212,27 @@ int chaseat(int dir_fd, const char *path, ChaseFlags flags, char **ret_path, int
...
-
         /* If a positive directory file descriptor is provided, always resolve the given path relative to it,
          * regardless of whether it is absolute or not. If we get AT_FDCWD, follow regular openat()
          * semantics, if the path is relative, resolve against the current working directory. Otherwise,
          * resolve against root. */
-        if (dir_fd >= 0 || !path_is_absolute(path))
-                fd = openat(dir_fd, ".", O_CLOEXEC|O_DIRECTORY|O_PATH);
-        else
-                fd = open("/", O_CLOEXEC|O_DIRECTORY|O_PATH);
+        fd = openat(dir_fd, done ?: ".", O_CLOEXEC|O_DIRECTORY|O_PATH);
```
Comment (daandemeyer) [inline src/basic/chase.c:230]: I wouldn't say that this is just refactoring, we now apply `dir_fd >= 0 && dir_fd_is_root(dir_fd) > 0` to `fd` as well whereas before we did not (which I think is a good thing to do).

**Review by DaanDeMeyer — APPROVED** [review]

File: src/basic/chase.c (hunk 1 above)
Comment (yuwata) [reply src/basic/chase.c:230]: The behavior is actually not changed. The commit message is updated. Please take a look.

**Comment by DaanDeMeyer** [comment]
We should probably check in chase() that the result of dir_fd_is_root() matches the result of empty_or_root() no? Otherwise we might assert() again if dir_fd_is_root() starts failing for other reasons again

**Comment by yuwata** [comment]
> We should probably check in chase() that the result of dir_fd_is_root() matches the result of empty_or_root() no? Otherwise we might assert() again if dir_fd_is_root() starts failing for other reasons again

If `path_is_root()` in `empty_or_root_to_null()` failed, then `dir_fd` may point root, but in `chaseat()`, we call `dir_fd_is_root()` for the fd, and that should be failed again, thus relative path should be returned from `chaseat()`...

But, yeah, for safety, we may need to propagate all failures in `dir_fd_is_root()` or `path_is_root()`. Will update.

(BTW, I think this PR is not necessary to fix the original issue. Let's clean up after the next release.)

**Comment by yuwata** [comment]
> > We should probably check in chase() that the result of dir_fd_is_root() matches the result of empty_or_root() no? Otherwise we might assert() again if dir_fd_is_root() starts failing for other reasons again
> 
> If `path_is_root()` in `empty_or_root_to_null()` failed, then `dir_fd` may point root, but in `chaseat()`, we call `dir_fd_is_root()` for the fd, and that should be failed again, thus relative path should be returned from `chaseat()`...
> 
> But, yeah, for safety, we may need to propagate all failures in `dir_fd_is_root()` or `path_is_root()`. I will update.
> 
> (BTW, I think this PR is not necessary to fix the original issue. Let's clean up after the next release.)

Done. @DaanDeMeyer PTAL.

**Comment by keszybz** [comment]
@DaanDeMeyer can you do another round? You've been working on this code more than others in recent times.

**Review by DaanDeMeyer — APPROVED** [review]

**Pull request merged**